### ISSUE TITLE
Go packages driver sometimes doesn't work with test files

### ISSUE THREAD

**Issue by csstaub** [bazelbuild/rules_go#2948]
I'm having an issue getting the updated Go packages driver working for some, but not other, test files. I've created a minimal repro to trigger the issue which appears to be related to dependencies of libraries not appearing on test targets. More info inline.

### What version of rules_go are you using?
70b8365a90e226b98f66bc35d3abb3e335883d81

### What version of gazelle are you using?
0.22.3

### What version of Bazel are you using?
4.2.1

### Does this issue reproduce with the latest releases of all the above?
Yes, this report is specifically about the latest Go packages driver.

### What operating system and processor architecture are you using?
macOS on x86-64

### Any other potentially useful information about your toolchain?
Nothing relevant

### What did you do?
I made a repo to reproduce this issue here: https://github.com/csstaub/gopackagesdriver-repro

Note that in the repro, `c:go_default_library` imports A/B, but the test target does not as the dependencies are not used in the tests explicitly. This builds fine because `c:go_default_test` has an `embed` with `c:go_default_library`, my understanding here is that the deps on the library hence become transitive deps of the test target.

But, if I run the Go packages driver for the test file, it reports no packages for `c_test.go`:
```
❯❯❯ echo '{}' | ./tools/gopackagesdriver.sh 'file=src/c/c_test.go'
INFO: Analyzed target @io_bazel_rules_go//go/tools/gopackagesdriver:gopackagesdriver (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
Target @io_bazel_rules_go//go/tools/gopackagesdriver:gopackagesdriver up-to-date:
  bazel-bin/external/io_bazel_rules_go/go/tools/gopackagesdriver/gopackagesdriver_/gopackagesdriver
INFO: Elapsed time: 0.280s, Critical Path: 0.00s
INFO: 1 process: 1 internal.
INFO: Build completed successfully, 1 total action
INFO: Build completed successfully, 1 total action
Running: [bazel info --tool_tag=gopackagesdriver --ui_actions_shown=0]
Running: [bazel query --tool_tag=gopackagesdriver --ui_actions_shown=0 --ui_event_filters=-info,-stderr --noshow_progress --order_output=no --output=label --nodep_deps --noimplicit_deps --notool_deps kind("go_library|go_test", same_pkg_direct_rdeps("src/c/c_test.go"))]
Running: [bazel build --tool_tag=gopackagesdriver --ui_actions_shown=0 --show_result=0 --build_event_json_file=/var/folders/p1/r8r4p94n281gfxs_p3jd38zh0000gn/T/gopackagesdriver_bep_95385647 --build_event_json_file_path_conversion=no --experimental_convenience_symlinks=ignore --ui_event_filters=-info,-stderr --noshow_progress --aspects=@io_bazel_rules_go//go/tools/gopackagesdriver:aspect.bzl%go_pkg_info_aspect --output_groups=go_pkg_driver_json_file,go_pkg_driver_stdlib_json_file,go_pkg_driver_srcs --keep_going //src/c:go_default_test]
ERROR: command succeeded, but not all targets were analyzed
{"NotHandled":false,"Sizes":{"WordSize":8,"MaxAlign":8},"Packages":[]}
```

Getting packages for `c.go` works fine:
```
❯❯❯ echo '{}' | ./tools/gopackagesdriver.sh 'file=src/c/c.go'
INFO: Analyzed target @io_bazel_rules_go//go/tools/gopackagesdriver:gopackagesdriver (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
Target @io_bazel_rules_go//go/tools/gopackagesdriver:gopackagesdriver up-to-date:
  bazel-bin/external/io_bazel_rules_go/go/tools/gopackagesdriver/gopackagesdriver_/gopackagesdriver
INFO: Elapsed time: 0.272s, Critical Path: 0.00s
INFO: 1 process: 1 internal.
INFO: Build completed successfully, 1 total action
INFO: Build completed successfully, 1 total action
Running: [bazel info --tool_tag=gopackagesdriver --ui_actions_shown=0]
Running: [bazel query --tool_tag=gopackagesdriver --ui_actions_shown=0 --ui_event_filters=-info,-stderr --noshow_progress --order_output=no --output=label --nodep_deps --noimplicit_deps --notool_deps kind("go_library|go_test", same_pkg_direct_rdeps("src/c/c.go"))]
Running: [bazel build --tool_tag=gopackagesdriver --ui_actions_shown=0 --show_result=0 --build_event_json_file=/var/folders/p1/r8r4p94n281gfxs_p3jd38zh0000gn/T/gopackagesdriver_bep_159641577 --build_event_json_file_path_conversion=no --experimental_convenience_symlinks=ignore --ui_event_filters=-info,-stderr --noshow_progress --aspects=@io_bazel_rules_go//go/tools/gopackagesdriver:aspect.bzl%go_pkg_info_aspect --output_groups=go_pkg_driver_json_file,go_pkg_driver_stdlib_json_file,go_pkg_driver_srcs --keep_going //src/c:go_default_library]
{"NotHandled":false,"Sizes":{"WordSize":8,"MaxAlign":8},"Roots":["//src/c:go_default_library"],"Packages":[{"ID":"//src/c:go_default_library","Name":"c","PkgPath":"example.com/c","GoFiles":["/Users/cstaub/Development/repro/src/c/c.go"],"CompiledGoFiles":["/Users/cstaub/Development/repro/src/c/c.go"],"ExportFile":"/private/var/tmp/_bazel_cstaub/821bede8a48c2229d6e91ced1c40f8a6/execroot/__main__/bazel-out/darwin-fastbuild/bin/src/c/go_default_library.x","Imports":{"example.com/a":"//src/a:go_default_library","example.com/b":"//src/b:go_default_library"}},{"ID":"//src/a:go_default_library","Name":"a","PkgPath":"example.com/a","GoFiles":["/Users/cstaub/Development/repro/src/a/a.go"],"CompiledGoFiles":["/Users/cstaub/Development/repro/src/a/a.go"],"ExportFile":"/private/var/tmp/_bazel_cstaub/821bede8a48c2229d6e91ced1c40f8a6/execroot/__main__/bazel-out/darwin-fastbuild/bin/src/a/go_default_library.x"},{"ID":"//src/b:go_default_library","Name":"b","PkgPath":"example.com/b","GoFiles":["/Users/cstaub/Development/repro/src/b/b.go"],"CompiledGoFiles":["/Users/cstaub/Development/repro/src/b/b.go"],"ExportFile":"/private/var/tmp/_bazel_cstaub/821bede8a48c2229d6e91ced1c40f8a6/execroot/__main__/bazel-out/darwin-fastbuild/bin/src/b/go_default_library.x"}]}
```

Not sure why this is -- I can get it to produce different outputs if I make the deps explicit, and/or add one but not both, to the test target. 

### What did you expect to see?
Package info for the test file.

### What did you see instead?
No package info for the test file.

**Comment by steeve** [comment]
yes, I am aware of this issue, which happens due to an error when the `.pkg.json` file is created at https://github.com/bazelbuild/rules_go/blob/master/go/tools/gopackagesdriver/aspect.bzl#L53-L56, more exactly during the call to `.to_json()`. I have tried to find the cause, so far no luck.

Basically, bazel will error out, but I don't know why...

**Comment by csstaub** [comment]
Is there a way to instrument this, to see the error(s)? Happy to help debug but I'm not sure how to go about it to get started.

**Comment by aignas** [comment]
An interesting thing is that if I add the following:
```starlark
    deps = [
        "//src/a:go_default_library",
        "//src/b:go_default_library",
    ],
```
to the provided repro examples `//src/c:go_default_test` dependencies, the driver succeeds.

I have also forked the example and tried reproducing the build with the latest available gazelle and rules_go versions and the issue dissapears if I change the package naming convention to the default one (https://github.com/aignas/gopackagesdriver-repro/commit/b67c336465e242fafdfcc88597b4904805f4345c). I have also played around adding various `print` messages in the funciton that @steeve mentioned above, but I did not observe any breaking of `to_json()` method.

Given the presence or absence of the faulty behaviour based on the BUILD file contents I would think that the issue is likely to be in the implementation of the aspect rule, but it is hard for me to tell because I am not familiar with how this works exactly.
